cpp/src/io/csv/reader_impl.cu

```diff
@@ -238,28 +232,41 @@ table_with_metadata reader::impl::read(size_t range_offset,
     const bool load_whole_file = range_offset == 0 && range_size == 0 && skip_rows <= 0 &&
                                  skip_end_rows <= 0 && num_rows == -1;
 
-    // Preload the intput data to device
-    if (load_whole_file) data_ = rmm::device_buffer(h_uncomp_data, h_uncomp_size);
-
-    // Pass nullptr for the device data is the data is not preloaded (will cause additional copies)
-    gather_row_offsets(
-      h_uncomp_data, h_uncomp_size, range_offset, stream, (load_whole_file ? &data_ : nullptr));
-
-    auto row_range = select_rows(
-      h_uncomp_data, h_uncomp_size, range_size, skip_rows, skip_end_rows, num_rows, stream);
-
-    data_size = row_range.second - row_range.first;
-    CUDF_EXPECTS(data_size <= h_uncomp_size, "Row range exceeds data size");
-
-    num_bits = (data_size + 63) / 64;
-    if (load_whole_file) {
-      // Loaded the whole file, add the start offset (e.g. empty rows) to the pointer
-      data_ptr = static_cast<char *>(data_.data()) + row_range.first;
-    } else {
-      // The start offset is applied to the device data buffer
-      data_    = rmm::device_buffer(h_uncomp_data + row_range.first, data_size);
-      data_ptr = static_cast<char *>(data_.data());
+    // With byte range, find the start of the first data row
+    size_t data_start_offset =
+      (range_offset != 0) ? find_first_row_start(h_uncomp_data, h_uncomp_size) : 0;
+
+    // FIXME: This is really a WAR for the datasource's inability to read bytes outside the mapped
+    // range. The header should be always parsed from the start of the file (would also allow
+    // access to csv column names with byte range)
+    size_t header_offset = (range_offset != 0) ? data_start_offset : 0;
+    size_t header_end =
+      parse_csv_header(h_uncomp_data + header_offset, h_uncomp_size - header_offset);
+
+    if (range_offset == 0) { data_start_offset = header_end; }
```

Comment:
You still need to call `parse_csv_header(h_uncomp_data + range_offset, ...)` when `range_offset != 0`.
As the comment says, it's a bit of a ugly hack that we parse the first data row as if it were the header (primarily needed to infer the number of columns). I tried to organize the code such that when we fix the datasource to allow reading the header outside the mapped range, we'll be able to just always parse the header at header_offset=0, and remove the various range_offset comparisons (rather than having two separate paths calling `parse_csv_header`).

Reply:
Oh, I see what you mean: `parse_csv_header` does indeed has a big side effect: it updates the `header` class member (something that was previously done in `select_rows`). Not ideal, but I kept this as is in an effort to limit the code changes (the `header` class member could probably just be a local in this function).

Reply:
Ah, I didn't notice that `parse_csv_header` and this scope are both methods of `reader::impl`

Hm, how about something like:
```cpp
size_t const header_offset = (range_offset != 0) ? find_first_row_start(h_uncomp_data, h_uncomp_size) : 0;
size_t const header_end = parse_csv_header(h_uncomp_data + header_offset, h_uncomp_size - header_offset);
size_t const data_start_offset = (range_offset != 0) ? header_offset : header_end;
```
or
```cpp
size_t const data_start_offset = [=]() {
  if (range_offset != 0) {
    size_t const offset = find_first_row_start(h_uncomp_data, h_uncomp_size);
    parse_csv_header(h_uncomp_data + offset, h_uncomp_size - header_offset);
    return offset;
  } else {
    return parse_csv_header(h_uncomp_data, h_uncomp_size);
  }
}();
```

Reply:
The top one works for me

Reply:
Done